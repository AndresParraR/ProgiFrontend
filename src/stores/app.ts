import { NotifyPosition } from "@/enums";
import type { AppState } from "@/types";
import { defineStore } from "pinia";
import type { QNotifyCreateOptions } from "quasar";

export const useAppStore = defineStore("app", {
  state: (): AppState => ({
    _notify: {
      message: "",
      type: "",
      timeout: 0,
      position: NotifyPosition.TOP_RIGHT,
    },
    _isLoading: false,
  }),
  getters: {
    notify(state) {
      return state._notify;
    },
    isLoading(state) {
      return state._isLoading;
    },
  },
  actions: {
    showNotification(
      message: string,
      type: string,
      timeout: number = 4000,
      position: QNotifyCreateOptions["position"] = NotifyPosition.TOP_RIGHT
    ) {
      this._notify = {
        message,
        type,
        timeout,
        position,
      };
    },
    setLoading(val: boolean) {
      this._isLoading = val;
    },
  },
});
