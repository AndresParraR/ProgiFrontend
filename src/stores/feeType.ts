import { FeeTypeService } from '@/services';
import type { FeeTypeState } from '@/types';
import { defineStore } from 'pinia';

export const useFeeTypeStore = defineStore('FeeType', {
  state: (): FeeTypeState => ({
    _feesType: [],
  }),
  getters: {
    feeType(state) {
      return state._feesType;
    },
  },
  actions: {
    async fetchVehiclesType() {
      const res = await FeeTypeService.getAll();
      this._feesType = res.data.response;
    },
  },
});
