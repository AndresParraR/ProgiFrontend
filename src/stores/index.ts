export * from "./app";
export * from "./vehicleTotal";
export * from "./vehicleType";
export * from "./feeType";
