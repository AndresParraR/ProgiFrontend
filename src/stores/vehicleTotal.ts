import { VehicleTotalService } from "@/services";
import type { CreateVehicleTotalParams, VehicleTotalState } from "@/types";
import { defineStore } from "pinia";

export const useVehiclesTotalStore = defineStore("vehiclesTotal", {
  state: (): VehicleTotalState => ({
    _vehiclesTotals: [],
  }),
  getters: {
    vehiclesTotal(state) {
      return state._vehiclesTotals;
    },
  },
  actions: {
    async fetchVehiclesTotal() {
      const res = await VehicleTotalService.getAll();
      this._vehiclesTotals = res.data.response;
    },
    async createVehiclesTotal(params: CreateVehicleTotalParams) {
      const res = await VehicleTotalService.create(params);
      this._vehiclesTotals.push(res.data.response);
    },
    async updateVehicleTotal(id: number, params: CreateVehicleTotalParams) {
      const res = await VehicleTotalService.update(id, params);
      this._vehiclesTotals[
        this._vehiclesTotals.findIndex((el) => el.id == id)
      ] = res.data.response;
    },
    async deleteVehicleTotal(id: number) {
      const res = await VehicleTotalService.delete(id);
      if (res)
        this._vehiclesTotals = this._vehiclesTotals.filter((el) => el.id != id);
    },
  },
});
