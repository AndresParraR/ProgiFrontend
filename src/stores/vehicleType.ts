import { VehicleTypeService } from '@/services';
import type { VehicleTypeState } from '@/types';
import { defineStore } from 'pinia';

export const useVehicleTypeStore = defineStore('vehicleType', {
  state: (): VehicleTypeState => ({
    _vehiclesType: [],
  }),
  getters: {
    vehicleType(state) {
      return state._vehiclesType;
    },
  },
  actions: {
    async fetchVehiclesType() {
      const res = await VehicleTypeService.getAll();
      this._vehiclesType = res.data.response;
    },
  },
});
