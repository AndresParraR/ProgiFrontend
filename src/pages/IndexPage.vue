<script setup lang="ts">
import { ref, type Ref } from 'vue';
//Components
import { FilterVehicle, TableVehicle, DialogVehicle } from '@/components';
import type { CreateVehicleTotalParams, VehicleTotal } from '@/types';
import { useAppStore, useVehiclesTotalStore } from '@/stores';

defineOptions({
  name: 'IndexPage',
});

const appStore = useAppStore();
const vehiclesTotalStore = useVehiclesTotalStore();

const search: Ref<string> = ref('');
const showDialogVehicle: Ref<boolean> = ref(false);
const vehicleTotal: Ref<VehicleTotal | null> = ref(null);

const handleSearch = (val: string) => {
  search.value = val;
};

const handleEditVehicleTotal = (val: VehicleTotal | null) => {
  vehicleTotal.value = val;
  showDialogVehicle.value = true;
};

const handleAddVehicle = () => {
  showDialogVehicle.value = true;
};

const addVehicle = async (val: CreateVehicleTotalParams) => {
  try {
    appStore.setLoading(true);
    await vehiclesTotalStore.createVehiclesTotal(val);
    appStore.showNotification('Vehicle succesfully created!', 'positive');
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    appStore.showNotification((error as any).message, 'negative');
  } finally {
    appStore.setLoading(false);
    closeDialogVehicle();
  }
};

const editVehicle = async (val: CreateVehicleTotalParams) => {
  try {
    appStore.setLoading(true);
    await vehiclesTotalStore.updateVehicleTotal(
      vehicleTotal.value?.id as number,
      val,
    );
    appStore.showNotification('Vehicle succesfully edited!', 'positive');
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    appStore.showNotification((error as any).message, 'negative');
  } finally {
    appStore.setLoading(false);
    closeDialogVehicle();
  }
};

const deleteVehicle = async (val: number) => {
  try {
    appStore.setLoading(true);
    await vehiclesTotalStore.deleteVehicleTotal(val);
    appStore.showNotification('Vehicle succesfully deleted!', 'positive');
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    appStore.showNotification((error as any).message, 'negative');
  } finally {
    appStore.setLoading(false);
  }
};

const closeDialogVehicle = () => {
  vehicleTotal.value = null;
  showDialogVehicle.value = false;
};
</script>

<template>
  <q-page class="row items-center justify-evenly">
    <q-card style="min-width: 900px">
      <DialogVehicle
        :show-dialog="showDialogVehicle"
        :vehicle="vehicleTotal!"
        @vehicle-added="($event: any) => addVehicle($event)"
        @vehicle-edited="($event: any) => editVehicle($event)"
        @close="closeDialogVehicle"
      />
      <FilterVehicle
        @search-updated="($event) => handleSearch($event)"
        @add-vehicle="() => handleAddVehicle()"
      />
      <TableVehicle
        :search="search"
        @edit-vehicle="handleEditVehicleTotal"
        @delete-vehicle="deleteVehicle"
      />
    </q-card>
  </q-page>
</template>
