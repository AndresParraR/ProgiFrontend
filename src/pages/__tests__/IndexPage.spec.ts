import { mount } from "@vue/test-utils";
import { IndexPage } from "@/pages";
import { beforeEach, describe, expect, test, vitest } from "vitest";
import { Quasar } from "quasar";
import { createTestingPinia } from "@pinia/testing";
import { useVehiclesTotalStore } from "@/stores";

describe("IndexPage Component", () => {
  let wrapper: any = null;

  beforeEach(() => {
    const global = {
      plugins: [
        createTestingPinia({
          createSpy: vitest.fn,
        }),
        Quasar,
      ],
    };

    wrapper = mount(IndexPage, {
      global,
    });
  });

  test("Should render correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test("should handleSearch change search value", async () => {
    const val = "text";

    wrapper.vm.handleSearch(val);

    expect(wrapper.vm.search).toEqual(val);
  });

  test("should handleEditVehicleTotal change vehicle value", async () => {
    const val = {
      id: 1,
      vehiclePrice: 501,
      vehicleTypeId: 1,
      vehicleTypeName: "Common",
      basicFee: 50,
      specialFee: 10.02,
      associationFee: 10,
      storageFee: 100,
      total: 671.02,
    };

    wrapper.vm.handleEditVehicleTotal(val);

    expect(wrapper.vm.vehicleTotal).toEqual(val);
    expect(wrapper.vm.showDialogVehicle).toEqual(true);
  });

  test("should handleAddVehicle change showDialogVehicle value", async () => {
    wrapper.vm.handleAddVehicle();

    expect(wrapper.vm.showDialogVehicle).toEqual(true);
  });

  test("should addVehicle", async () => {
    const vehiclesTotalStore = useVehiclesTotalStore();

    wrapper.vm.vehicleTotal = {
      id: 1,
      vehiclePrice: 501,
      vehicleTypeId: 1,
      vehicleTypeName: "Common",
      basicFee: 50,
      specialFee: 10.02,
      associationFee: 10,
      storageFee: 100,
      total: 671.02,
    };

    wrapper.vm.showDialogVehicle = true;

    await wrapper.vm.addVehicle();

    expect(vehiclesTotalStore.createVehiclesTotal).toBeCalled();
    expect(wrapper.vm.vehicleTotal).toEqual(null);
    expect(wrapper.vm.showDialogVehicle).toEqual(false);
  });

  test("should editVehicle", async () => {
    const vehiclesTotalStore = useVehiclesTotalStore();

    wrapper.vm.vehicleTotal = {
      id: 1,
      vehiclePrice: 501,
      vehicleTypeId: 1,
      vehicleTypeName: "Common",
      basicFee: 50,
      specialFee: 10.02,
      associationFee: 10,
      storageFee: 100,
      total: 671.02,
    };

    wrapper.vm.showDialogVehicle = true;

    await wrapper.vm.editVehicle();

    expect(vehiclesTotalStore.updateVehicleTotal).toBeCalled();
    expect(wrapper.vm.vehicleTotal).toEqual(null);
    expect(wrapper.vm.showDialogVehicle).toEqual(false);
  });

  test("should deleteVehicle", async () => {
    const vehiclesTotalStore = useVehiclesTotalStore();

    await wrapper.vm.deleteVehicle();

    expect(vehiclesTotalStore.deleteVehicleTotal).toBeCalled();
  });
});
