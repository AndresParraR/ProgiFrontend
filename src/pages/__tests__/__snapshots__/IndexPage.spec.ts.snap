// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IndexPage Component > Should render correctly 1`] = `<!---->`;
